lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -613,6 +619,106 @@ void TypeLoweringVisitor::visitStmt(WhenOp op) {
   lowerBlock(&op.getElseBlock());
 }
 
+// Convert an aggregate type into a flat list of fields.  This is used
+// when working with instances and mems to flatten them.
+static void flattenType(FIRRTLType type, SmallVectorImpl<FIRRTLType> &results) {
+  std::function<void(FIRRTLType)> flatten = [&](FIRRTLType type) {
+    TypeSwitch<FIRRTLType>(type)
+        .Case<BundleType>([&](auto bundle) {
+          for (auto &elt : bundle.getElements())
+            flatten(elt.type);
+          return;
+        })
+        .Case<FVectorType>([&](auto vector) {
+          for (size_t i = 0, e = vector.getNumElements(); i != e; ++i)
+            flatten(vector.getElementType());
+          return;
+        })
+        .Default([&](auto) {
+          results.push_back({type});
+          return;
+        });
+  };
+  return flatten(type);
+}
+
+// Create subfield and subaccess to access all the leaf elements of the result
+// of the input op and insert the results into the flatData.
+void TypeLoweringVisitor::crerateAggregateAccess(
+    SubfieldOp op, SmallVectorImpl<Value> &flatData) {
+  std::function<void(Value val)> flatten = [&](Value val) {
+    TypeSwitch<FIRRTLType>(val.getType().cast<FIRRTLType>())
+        .Case<BundleType>([&](BundleType bundle) {
+          for (auto &elt : llvm::enumerate(bundle.getElements())) {
+            auto subF = builder->create<SubfieldOp>(val, elt.index());
+            flatten(subF.getResult());
+          }
+          return;
+        })
+        .Case<FVectorType>([&](auto vector) {
+          for (size_t i = 0, e = vector.getNumElements(); i != e; ++i) {
+            auto subF = builder->create<SubindexOp>(val, i);
+            flatten(subF.getResult());
+          }
+          return;
+        })
+        .Default([&](auto) {
+          flatData.push_back(val);
+          return;
+        });
+  };
+  return flatten(op.getResult());
+}
+
+/// Map the flattened memory ports to the original memory.
+void TypeLoweringVisitor::flattenMem(SubfieldOp oldField, StringRef name,
+                                     SmallVectorImpl<unsigned> &maskBits,
+                                     Value newMemRes, unsigned fieldIndex,
+                                     bool isRead) {
+  SmallVector<Value, 8> flatData;
+  crerateAggregateAccess(oldField, flatData);
+  if (name == "mask" || name == "wmask") {
+    SmallVector<Value, 8> tmp;
+    for (auto maskB : llvm::enumerate(maskBits)) {
+      auto m = flatData[maskB.index()];
+      for (unsigned repeat = 0; repeat < maskB.value(); ++repeat)
+        tmp.push_back(m);
+    }
+    flatData = tmp;
+  }
+  auto newData = getSubWhatever(newMemRes, fieldIndex);
+  size_t uptoBits = 0;
+  WireOp lastIndexWire;
+
+  for (auto fData : flatData) {
+    auto fieldBits = fData.getType().cast<FIRRTLType>().getBitWidthOrSentinel();
```

Comment:
Can this all be simplified with a struct/vector -> int cast instead?

Reply:
Updated to IntType